### ISSUE TITLE
Doubts about model inference

### ISSUE THREAD

**Issue by eanson023** [sunnweiwei/GenRet#2]
Hi Wei,

Hello author, your work is very fascinating. I've been reading your code for the past two days. I have always had some questions about the inference (test) function. I don’t know how you bind `query` to `doc`. I see that you called the `model.generate` function, but the model did not use some language modeling objectives during training. Specifically, this is the code below. Can you explain it to me? (I really don’t understand it, lol)

![image](https://github.com/sunnweiwei/GenRet/assets/57490022/9a29f38f-1772-4ae6-839b-bd47c4208c4b)
![image](https://github.com/sunnweiwei/GenRet/assets/57490022/9543af3a-f21c-4bb2-885d-9dfd356409f8)

Also, according to my understanding, Codebook does not seem to play a role in inferencing. I have not yet found any code that uses Codebook for inference. (My GPU only has 16g memory, so it is difficult to fully reproduce GenRET)

Best Regards,

Sheng Yan
2024-1-13

**Comment by sunnweiwei** [comment]
Hi,

The model is trained with language modeling objectives. For example, see : `query_ce_loss = F.cross_entropy(query_prob, codes_doc.detach())`, where `query_prob` is the generation probability and `codes_doc` is the label.

The codebook is used during inference stage: the `generate` function calls the `forward` function within its generation loop, where the codebook `self.centroids` is used.

**Comment by eanson023** [comment]
Hi @sunnweiwei , Thank you for your prompt reply, I understand what you mean.  `huggingface` encapsulates the function so well that I didn't understand it for a while.  I'll take a look tomorrow.  

Another question I have is that since the `query` is mapped to the virtual (generated) `docid`, how to calculate the recall with the real `doc`?  Does the real doc also need to be converted to `docid`?  I didn’t understand this in the code either, haha

**Comment by ntuwork** [comment]
> Hi @sunnweiwei , Thank you for your prompt reply, I understand what you mean. `huggingface` encapsulates the function so well that I didn't understand it for a while. I'll take a look tomorrow.
> 
> Another question I have is that since the `query` is mapped to the virtual (generated) `docid`, how to calculate the recall with the real `doc`? Does the real doc also need to be converted to `docid`? I didn’t understand this in the code either, haha

Yes, I also have the same question.

```query_ce_loss = F.cross_entropy(query_prob, codes_doc.detach())```

In this step, why is it possible to use ```codes_doc``` as a label? ```codes_doc``` I understand that it seems to need to be further optimized according to the training fault. Because I understand docid is obtained based on the reconstruction of documents training. I hope you can help me explain, thanks!

**Comment by eanson023** [comment]
Hi @ntuwork , Here's my understanding of this code (I don't know if it's correct):

1. Using this codebook index for labeling is indeed a dynamic process. For a certain piece of data, the label may be `1` in the first iteration, but change to another index in the next iteration. The author talked about this challenge in Section 3.2, so it proposed a progressive training scheme (Section 3.2.1) to alleviate this problem.

2. Regarding reconstruction loss, this should only be an auxiliary objective. It feels like it can still work without it, but the performance may just be a little worse.

**Comment by sunnweiwei** [comment]
Hi. Yes during training, the model uses an on-the-fly docid as the target in the commitment loss, even though it may be sub-optimal and vary across different epochs. We monitor the model's convergence by comparing the consistency of docids between adjacent epochs. We have found that the model can achieve good convergence (e.g., 95% id consistency) when the codebook is properly initialized.

**Comment by eanson023** [comment]
Hi @sunnweiwei , Thank you very much for your reply. This on-the-fly docid assignment is indeed very clever and can achieve end-to-end training effects. In the past few days, I have applied your ideas to my research. In my experiments, I found that most of the data will be assigned to the **same docid**, resulting in high test results, but it is undesirable.

I would like to ask you, when I did this experiment, my codebook did not use the `warm up` training method mentioned in the paper. Is it because of this that many data will be assigned to the same docid? In other words, the codebook with a long enough time step mentioned in the paper can guarantee the uniqueness of docid assignment does not seem to be reflected. I don’t know if it’s because I didn’t use `progressive training techniques` and `warm up` training methods.

**Comment by sunnweiwei** [comment]
Hi @eanson023 

> Another question I have is that since the query is mapped to the virtual (generated) docid, how to calculate the recall with the real doc? Does the real doc also need to be converted to docid? I didn’t understand this in the code either, haha

I build a dict `docid_to_doc` that map virtual docid to real doc (`string -> list`). Then for a model generated virtual docid, I map it to real doc and calculate recall using real doc.

I build a dict `docid_to_doc` that maps virtual docids to real docs (`string -> list`). Then, for a model-generated virtual docid, I map it to the real docs and calculate recall using real docs.

> I would like to ask you, when I did this experiment, my codebook did not use the warm up training method mentioned in the paper. Is it because of this that many data will be assigned to the same docid? In other words, the codebook with a long enough time step mentioned in the paper can guarantee the uniqueness of docid assignment does not seem to be reflected. I don’t know if it’s because I didn’t use progressive training techniques and warm up training methods.

I think warm-up is necessary to ensure diverse docid assignment. Based on my experience, to achieve a balanced distribution of ID assignments, the initial ID distribution (the distribution before training starts) is important. I suspect this is because if an ID is rarely selected initially, it will receive limited positive gradient, which might further prevent it from being selected. Therefore, to ensure a balanced initialization, I initialize the codebook parameters using cluster centroids (from constrained k-means).

imo longer docids can accommodate more unique items, when the distribution of each step is relatively uniform. Otherwise, for example, if all items are assigned the same id at each step, increasing the length will not increase capacity.

**Comment by ntuwork** [comment]
> Hi @eanson023
> 
> > Another question I have is that since the query is mapped to the virtual (generated) docid, how to calculate the recall with the real doc? Does the real doc also need to be converted to docid? I didn’t understand this in the code either, haha
> 
> I build a dict `docid_to_doc` that map virtual docid to real doc (`string -> list`). Then for a model generated virtual docid, I map it to real doc and calculate recall using real doc.
> 
> I build a dict `docid_to_doc` that maps virtual docids to real docs (`string -> list`). Then, for a model-generated virtual docid, I map it to the real docs and calculate recall using real docs.
> 
> > I would like to ask you, when I did this experiment, my codebook did not use the warm up training method mentioned in the paper. Is it because of this that many data will be assigned to the same docid? In other words, the codebook with a long enough time step mentioned in the paper can guarantee the uniqueness of docid assignment does not seem to be reflected. I don’t know if it’s because I didn’t use progressive training techniques and warm up training methods.
> 
> I think warm-up is necessary to ensure diverse docid assignment. Based on my experience, to achieve a balanced distribution of ID assignments, the initial ID distribution (the distribution before training starts) is important. I suspect this is because if an ID is rarely selected initially, it will receive limited positive gradient, which might further prevent it from being selected. Therefore, to ensure a balanced initialization, I initialize the codebook parameters using cluster centroids (from constrained k-means).
> 
> imo longer docids can accommodate more unique items, when the distribution of each step is relatively uniform. Otherwise, for example, if all items are assigned the same id at each step, increasing the length will not increase capacity.

I would like to ask that Does the provided code perform a warmup implementation? Because in the first round of training I noticed that IDs are also generated for docs.

Thanks!

**Comment by sunnweiwei** [comment]
Warm up is included in the code, which essentially uses clustering centroids of doc embeddings to initialize the codebook.
